core/tests/puppeteer-acceptance-tests/functions/is-element-clickable.ts

```diff
@@ -0,0 +1,170 @@
+  };
+
+  /**
+   * This function combines the overlapping element and overlapping rect
+   * to get the element or elements that are overlapping the given
+   * element.
+   */
+  const getOverlappingElements = (
+    element: Element,
+    parent: ShadowRoot | Document = document
+  ): Element[] => {
+    const overlappingElement = getOverlappingElement(element, parent);
+    const overlappingRect = getOverlappingRect(element, parent);
+    const overlappingElements: Element[] = [];
+    if (overlappingElement) {
+      overlappingElements.push(overlappingElement);
+    }
+    if (overlappingRect) {
+      overlappingElements.push(overlappingRect);
+    }
+    return overlappingElements;
+  };
+
+  /**
+   * This function is recursive and checks if the target element is the only
+   * overlapping element meaning there are no other elements blocking
+   * the target element.
+   *
+   * @param {Element} targetElement - The element to compare with the other
+   * overlapping elements.
+   * @param {Element[]} overlappingElements - The overlapping elements to compare
+   * against the target element.
+   */
+  const isOverlappingElementsMatching = (
+    targetElement: Element,
+    overlappingElements: Element[]
+  ): boolean => {
+    if (
+      overlappingElements.some(
+        el => el === targetElement || targetElement.contains(el)
+      )
+    ) {
+      return true;
+    }
+
+    let elementsWithShadow = [...new Set(overlappingElements)];
+    elementsWithShadow = elementsWithShadow.filter(
+      el => el && el.shadowRoot && el.shadowRoot.elementFromPoint
+    );
+
+    let shadowElements: Element[] = [];
+    for (const shadowEl of elementsWithShadow) {
+      if (shadowEl.shadowRoot) {
+        shadowElements.push(
+          ...getOverlappingElements(targetElement, shadowEl.shadowRoot)
+        );
+      }
+    }
+    shadowElements = [...new Set(shadowElements)];
+    shadowElements = shadowElements.filter(
+      el => !overlappingElements.includes(el)
+    );
+
+    if (shadowElements.length === 0) {
+      return false;
+    }
+
+    return isOverlappingElementsMatching(targetElement, shadowElements);
+  };
+
+  /**
+   * This function checks if the given element is in the viewport.
+   */
+  const isElementInViewport = (element: Element): boolean => {
+    const elementDimensions = element.getBoundingClientRect();
+
+    const windowHeight =
+      window.innerHeight || document.documentElement.clientHeight;
+    const windowWidth =
+      window.innerWidth || document.documentElement.clientWidth;
+
+    const verticalInView =
+      elementDimensions.top <= windowHeight &&
+      elementDimensions.top + elementDimensions.height > 0;
+    const horizontalInView =
+      elementDimensions.left <= windowWidth &&
+      elementDimensions.left + elementDimensions.width > 0;
+
+    return verticalInView && horizontalInView;
+  };
+
+  /**
+   * This function checks if the element is clickable, by checking if it is
```

Comment:
I think this (and probably other docstrings) might be better written as jsdoc with documentation for `@param` and `@returns`. That way you can define the return value more crisply: "Whether the element is clickable -- i.e. in the viewport, not blocked by any other element, and not disabled.